resource_estimator/examples/basic_qsharp.rs

```diff
@@ -0,0 +1,94 @@
+#![allow(clippy::cast_precision_loss)]
+
+// This example illustrates how to use the Resource Estimator API to perform a
+// standard resource estimation on top of the qubit and QEC models provided by
+// the systems architecture that we ship. We are running resource estimation on
+// the Dynamics sample that is in the samples directory of the repository.
+
+use std::{fs::read_to_string, io, rc::Rc};
+
+use qsc::{
+    compile::{core, std},
+    interpret::{GenericReceiver, Interpreter},
+    LanguageFeatures, PackageStore, PackageType, SourceMap, TargetCapabilityFlags,
+};
+use resource_estimator::{
+    counts::LogicalCounter,
+    estimates::{ErrorBudget, PhysicalResourceEstimation},
+    system::{floquet_code, LogicalResourceCounts, PhysicalQubit, TFactoryBuilder},
+};
+
+fn main() {
+    // The sample is equivalent to the simpler sample `basic_logical_counts.rs`,
+    // which we advise to check first. We are using here the same error
+    // correction code, the same qubit model, the same factory builder, and the
+    // same error budget. The difference is that we are computing the logical
+    // post-layout overhead from executing a Q# file.
+
+    // 1) A quantum error correction code
+    let code = floquet_code();
+
+    // 2) A qubit model
+    let qubit = Rc::new(PhysicalQubit::qubit_maj_ns_e6());
+
+    // 3) A factory builder to provide magic states
+    let builder = TFactoryBuilder::default();
+
+    // 4) The logical resource overhead; in this example we are evaluating a Q#
+    //    file to retrieve the logical resource counts.
+    let logical_counts = Rc::new(compute_logical_counts_from_file());
+
+    // 5) An error budget
+    let budget = ErrorBudget::from_uniform(0.001);
+
+    // Perform resource estimation and print main result metrics
+    let estimation = PhysicalResourceEstimation::new(code, qubit, builder, logical_counts, budget);
+    let result = estimation.estimate().expect("estimation does not fail");
+
+    println!("Number of physical qubits: {}", result.physical_qubits());
+    println!(
+        "Runtime:                   {:.2e} secs",
+        result.runtime() as f64 / 1e9
+    );
+}
+
+fn compute_logical_counts_from_file() -> LogicalResourceCounts {
```

Comment:
- Given that interpreter and Package store objects might be unfamiliar to some users, it might be helpful to provide a more abstract layer. Thoughts?

- If we decide to abstract this, abstracting it at `qsharp.json` level might also help. Thoughts?

Reply:
Thanks for your feedback. In fact, I don't want to highlight the interpreter mechanism in this sample, but rather provide a sample that can easily be customized. I don't expect that this part of the code will change, except for maybe the source files. But explaining that multiple sources could be loaded can help.

Can you elaborate on the second point please. I don't quite understand it.

Reply:
> Thanks for your feedback. In fact, I don't want to highlight the interpreter mechanism in this sample, but rather provide a sample that can easily be customized. I don't expect that this part of the code will change, except for maybe the source files. But explaining that multiple sources could be loaded can help.

I feel if we have an API along the lines of  `compute_logical_counts_from_file(file, estimate_op)`, it simplifies user experience greatly since many might not familiar with internal workings. It seems the intension is that users will change the file_name and the estimate_operation, which we could simplify by having these things as arguments. Also, if we have this function in this example, the user might be more tempted to dig into this logic. Maybe this functionality can be provided via `resource-estimator` module.

> Can you elaborate on the second point please. I don't quite understand it.

For the second point, the intension is to be able to have an API that takes folder containing path to qsharp.json, or a valid path including `qsharp.json`. Eg: `samples\estimation\df-chemistry\qsharp.json` OR `samples\estimation\df-chemistry\`

Reply:
Thanks for the clarification. Meanwhile, I updated the PR to not contain the Q# sample and to only focus on the Rust functions in the `resource_estimator` crate. Thanks for your feedback!